Framework/Algorithms/src/ReflectometryBackgroundSubtraction.cpp

```diff
@@ -0,0 +1,196 @@
+
+void ReflectometryBackgroundSubtraction::calculateAverageSpectrumBackground(
+    API::MatrixWorkspace_sptr inputWS) {
+
+  auto alg = createChildAlgorithm("AverageSpectrumBackground");
+  alg->initialize();
+  alg->setProperty("InputWorkspace", inputWS);
+  alg->setProperty("BottomBackgroundRange",
+                   getPropertyValue("BottomBackgroundRange"));
+  alg->setProperty("TopBackgroundRange",
+                   getPropertyValue("TopBackgroundRange"));
+  alg->execute();
+  API::MatrixWorkspace_sptr output = alg->getProperty("OutputWorkspace");
+  setProperty("OutputWorkspace", output);
+}
+
+void ReflectometryBackgroundSubtraction::calculatePolynomialBackground(
+    API::MatrixWorkspace_sptr inputWS) {
+  // if the input workspace is an event workspace it must be converted to a
+  // Matrix workspace as cannot transpose a event workspace
+  // if it is already a MatrixWorkspace this does nothing
+  auto convert = createChildAlgorithm("ConvertToMatrixWorkspace");
+  convert->setProperty("InputWorkspace", inputWS);
+  convert->execute();
+  API::MatrixWorkspace_sptr convertedWS =
+      convert->getProperty("OutputWorkspace");
+
+  // transpose the workspace before the background can be found
+  auto transpose = createChildAlgorithm("Transpose");
+  transpose->setProperty("InputWorkspace", convertedWS);
+  transpose->execute();
+  API::MatrixWorkspace_sptr transposedWS =
+      transpose->getProperty("OutputWorkspace");
+
+  auto poly = createChildAlgorithm("CalculatePolynomialBackground");
+  poly->initialize();
+  poly->setProperty("InputWorkspace", transposedWS);
+  poly->setProperty("Degree", getPropertyValue("DegreeOfPolynomial"));
+  poly->setProperty("XRanges", getPropertyValue("XRanges"));
+  poly->setProperty("CostFunction", getPropertyValue("CostFunction"));
+  poly->execute();
+  API::MatrixWorkspace_sptr bgd = poly->getProperty("OutputWorkspace");
+
+  // transpose again to get the background
+  transpose->setProperty("InputWorkspace", bgd);
+  transpose->execute();
+  API::MatrixWorkspace_sptr transposedBgd =
+      transpose->getProperty("OutputWorkspace");
+
+  // subtract the background from the input workspace
+  auto output = minus(inputWS, transposedBgd);
+  setProperty("OutputWorkspace", output);
+}
+
+//----------------------------------------------------------------------------------------------
+/** Initialize the algorithm's properties.
+ */
+void ReflectometryBackgroundSubtraction::init() {
+
+  // Input workspace
+  declareProperty(make_unique<WorkspaceProperty<MatrixWorkspace>>(
+                      "InputWorkspace", "", Direction::Input,
+                      boost::make_shared<CommonBinsValidator>()),
+                  "An input workspace.");
+
+  std::vector<std::string> backgroundTypes = {"Per Detector Average",
+                                              "Polynomial"};
+  declareProperty("TypeOfBackgroundSubtraction", "Per Detector Average",
```

Comment:
It feels like we could have a shorter name here, perhaps `SubtractionMethod`? I'm not sure if there's a naming convention, but that would be in keeping with `CorrectionMethod` used in `PolarizationEfficiencyCor`